eval/topdown.go

```diff
@@ -540,6 +588,174 @@ func evalRefRec(ctx *TopDownContext, ref opalog.Ref, iter TopDownIterator, path
+
+	return TopDown(child, func(child *TopDownContext) error {
+		value := child.Bindings.Get(rule.Value.Value)
+		if value == nil {
+			return fmt.Errorf("unbound variable: %v", rule.Value)
+		}
+		return evalRefRuleResult(ctx, ref, ref[len(path)+1:], value, iter)
+	})
+}
+
+func evalRefRulePartialSetDoc(ctx *TopDownContext, ref opalog.Ref, path opalog.Ref, rule *opalog.Rule, iter TopDownIterator) error {
+	suffix := ref[len(path):]
+	if len(suffix) == 0 {
+		return fmt.Errorf("not implemented: %v %v %v", ref, path, rule)
+	}
+
+	if len(suffix) > 1 {
+		// TODO(tsandall): attempting to dereference set lookup
+		// results in undefined value, catch this using static analysis
+		return nil
+	}
+
+	if !suffix[0].IsGround() {
+		child := ctx.Child(rule, newHashMap())
+		return TopDown(child, func(child *TopDownContext) error {
+			value := child.Bindings.Get(rule.Key.Value)
+			if value == nil {
+				return fmt.Errorf("unbound variable: %v", rule.Key)
+			}
+			// Take the output of the child context and bind (1) the value to
+			// the variable from this context and (2) the reference to true
+			// so that expression will be defined. E.g., given a simple rule:
+			// "p = true :- q[x]", we say that "p" should be defined if "q"
+			// is defined for some value "x".
+			ctx = ctx.BindVar(suffix[0].Value.(opalog.Var), value)
+			ctx = ctx.BindRef(ref[:len(path)+1], opalog.Boolean(true))
+			return iter(ctx)
+		})
+	}
+
+	bindings := newHashMap()
+	bindings.Put(rule.Key.Value, suffix[0].Value)
+	child := ctx.Child(rule, bindings)
+	return TopDown(child, func(child *TopDownContext) error {
+		// See comment above for explanation of why the reference is bound to true.
+		ctx = ctx.BindRef(ref[:len(path)+1], opalog.Boolean(true))
+		return iter(ctx)
+	})
+
+}
+
+func evalRefRuleResult(ctx *TopDownContext, ref opalog.Ref, suffix opalog.Ref, result opalog.Value, iter TopDownIterator) error {
+
+	switch result := result.(type) {
+	case opalog.Ref:
+		// Below we concatenate the result of evaluating a rule with the rest of the reference
+		// to be evaluated.
+		//
+		// E.g., given two rules: q[k] = v :- a[k] = v and p = true :- q[k].a[i] = 100
+		// when evaluating the expression in "p", this function will be called with a binding for
+		// "q[k]" and "k". This leaves the remaining part of the reference (i.e., ["a"][i])
+		// needing to be processed. This is done by substituting the prefix of the original reference
+		// with the binding. In this case, the prefix is "q[k]" and the binding value would be
+		// "a[<some key>]".
+		var binding opalog.Ref
+		binding = append(binding, result...)
+		binding = append(binding, suffix...)
+		return evalRefRec(ctx, suffix, func(ctx *TopDownContext) error {
+			ctx = ctx.BindRef(ref, plugValue(binding, ctx.Bindings))
+			return iter(ctx)
+		}, result)
+
+	case opalog.Array:
+		if len(suffix) > 0 {
+			return result.Query(suffix, func(keys map[opalog.Var]opalog.Value, value opalog.Value) error {
```

Comment:
There's an issue here which is similar to the one above from evalRefRulePartialSetDoc. I.e., the bindings need to be applied to the suffix before running querying the array (or object) below.